src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicAnalysisDataReader.cs

```diff
@@ -0,0 +1,367 @@
+    {
+        public ImmutableArray<DynamicAnalysisDocument> Documents { get; }
+        public ImmutableArray<DynamicAnalysisMethod> Methods { get; }
+
+        private readonly Blob _guidHeapBlob;
+        private readonly Blob _blobHeapBlob;
+
+        private const int GuidSize = 16;
+
+        public DynamicAnalysisDataReader(byte* buffer, int size)
+        {
+            var reader = new BlobReader(buffer, size);
+
+            // header:
+            if (reader.ReadByte() != 'D' || reader.ReadByte() != 'A' || reader.ReadByte() != 'M' || reader.ReadByte() != 'D')
+            {
+                throw new BadImageFormatException();
+            }
+
+            // version
+            byte major = reader.ReadByte();
+            byte minor = reader.ReadByte();
+            if (major != 0 && minor != 1)
+            {
+                throw new NotSupportedException();
+            }
+
+            // table sizes:
+            int documentRowCount = reader.ReadInt32();
+            int methodSpanRowCount = reader.ReadInt32();
+
+            // blob heap sizes:
+            int stringHeapSize = reader.ReadInt32();
+            int userStringHeapSize = reader.ReadInt32();
+            int guidHeapSize = reader.ReadInt32();
+            int blobHeapSize = reader.ReadInt32();
+
+            // TODO: check size ranges
+
+            bool isBlobHeapSmall = blobHeapSize < 0x10000;
+            bool isGuidHeapSmall = guidHeapSize / GuidSize < 0x10000; // TODO
+
+            var documentsBuilder = ArrayBuilder<DynamicAnalysisDocument>.GetInstance(documentRowCount);
+
+            for (int i = 0; i < documentRowCount; i++)
+            {
+                var name = MetadataTokens.BlobHandle(ReadReference(ref reader, isBlobHeapSmall));
+                var hashAlgorithm = MetadataTokens.GuidHandle(ReadReference(ref reader, isGuidHeapSmall));
+                var hash = MetadataTokens.BlobHandle(ReadReference(ref reader, isBlobHeapSmall));
+
+                documentsBuilder.Add(new DynamicAnalysisDocument(name, hashAlgorithm, hash));
+            }
+
+            Documents = documentsBuilder.ToImmutableAndFree();
+
+            var methodsBuilder = ArrayBuilder<DynamicAnalysisMethod>.GetInstance(methodSpanRowCount);
+
+            for (int i = 0; i < methodSpanRowCount; i++)
+            {
+                methodsBuilder.Add(new DynamicAnalysisMethod(MetadataTokens.BlobHandle(ReadReference(ref reader, isBlobHeapSmall))));
+            }
+
+            Methods = methodsBuilder.ToImmutableAndFree();
+
+            int stringHeapOffset = reader.Offset;
+            int userStringHeapOffset = stringHeapOffset + stringHeapSize;
+            int guidHeapOffset = userStringHeapOffset + userStringHeapSize;
+            int blobHeapOffset = guidHeapOffset + guidHeapSize;
+
+            if (reader.Length != blobHeapOffset + blobHeapSize)
+            {
+                throw new BadImageFormatException();
```

Comment:
Since there are multiple places in this message where a `BadImageFormatException` can be thrown, would it be useful to give them distinguishing messages?